lib/core/include/irods/system_error.hpp

```diff
@@ -0,0 +1,80 @@
+#ifndef IRODS_SYSTEM_ERROR_HPP
+#define IRODS_SYSTEM_ERROR_HPP
+
+/// \file
+
+#include <compare>
+#include <system_error>
+
+namespace irods::experimental
+{
+    /// An error category that contains iRODS specific logic for dealing with
+    /// iRODS error codes.
+    ///
+    /// See https://cppreference.com for additional details about std::error_category.
+    ///
+    /// \since 4.2.12
+    class error_category : public std::error_category
+    {
+      public:
+        [[nodiscard]] auto name() const noexcept -> const char* override
+        {
+            return "iRODS";
+        } // name
+
+        [[nodiscard]] auto message(int _condition) const -> std::string override;
+
+        auto operator==(const error_category& _rhs) const noexcept -> bool
+        {
+            return this == &_rhs;
+        } // operator==
+
+        auto operator<=>(const error_category& _rhs) const noexcept -> std::strong_ordering
+        {
+            if (this < &_rhs) {
+                return std::strong_ordering::less;
+            }
+
+            if (this > &_rhs) {
+                return std::strong_ordering::greater;
+            }
+
+            return std::strong_ordering::equal;
+        } // operator<=>
+    }; // class error_category
+
+    /// Identifies the iRODS error category.
+    ///
+    /// Obtains a reference to the static error category object for iRODS errors. The object
+    /// is required to override the virtual function \p std::error_category::name() to return
+    /// a pointer to the string \p "iRODS". It is used to identify error conditions that
+    /// correspond to the iRODS error codes defined in rodsErrorTable.h.
+    ///
+    /// \return A reference to the static object derived from \p std::error_category.
+    ///
+    /// \since 4.2.12
+    [[nodiscard]] auto irods_category() noexcept -> const error_category&;
```

Comment:
Can we put the word "error" somewhere in the name to indicate what this is for? Right now, it will just show up as `irods::experimental::irods_category` which is vague to me.

Reply:
This is mirroring the scene used by the C++ std lib (e.g. `generic_category`, `system_category`, `future_category`, `iostream_category`). No one should be reaching for this in 99% of the cases.

Developers should prefer `make_error_code` instead of constructing `std::error_code` objects themselves. That function will call `irods_category` for you.

Does that make sense?

Reply:
Oooh, I did not know that. Okay, then I'm fine with this. I don't like it, but that's just how it is for these categories!